### DISCUSSION TITLE
Discussion about the whole project, Especially spring-data-jdbc and gradle

### DISCUSSION THREAD

**Discussion by ichenhe** [querydsl/querydsl#3038] [General]
First of all, there is no doubt that this is a great library, thanks to all the contributors. But after three days of struggle, I finally found that the structure of the project seemed too old. Here are my personal thoughts:

### 1.  The documents are a little messy.

The guide on the web page looks very organized, but it actually contains many outdated content. Or, some new features have not been added. I want to help improve it, but the document generator itself is old too. I am glad to hear that the maintainer has [a plan to reconstruct it](https://github.com/querydsl/querydsl/discussions/3036#discussioncomment-1367488).

### 2. Too many bindings with JPA

Maybe the title is not very clear. I mean, too many people use it with spring data JPA, but [they were forced to](https://github.com/querydsl/querydsl/issues/350#issuecomment-437333736). JPA is too bloated, especially the cascading is very chaotic. So, I believe that a lot of people want to escape.

Spring data jdbc (not `spring jdbc`) is a new product of the spring-data series. It is between jdbc and jpa and is a good substitute. Unfortunately, querydsl does not have out-of-the-box support for it.

### 3. Q class code gen is not consistent

As far as I understand, there are two ways of code generation at present.

1. Generate from annotation.
2. Connect the database and scan them.

Obviously, the second way is not very convenient. (And does not support gradle)  
The first method can only be used perfectly with JPA -- The maintainer said that without annotations from JPA, we would not be able to get enough metadata.

But `@Table`, `@Column` are  supported throughout Spring Data, not only Spring Data JPA. At the same time, querydsl has its own `@QueryEntity`. Does this mean that for spring-data-jdbc, we can achieve the same result as JPA?

### 4. Not friendly to gradle

First of all, APT can be used with gradle, or even simpler than maven. But only for JPA (see problem 3). 

However, the generation method based on database scanning [does not seem to be suitable for gradle projects (or very difficult)](https://github.com/querydsl/querydsl/issues/1976#issuecomment-235031357).

This means that a gradle project that does not use JPA will not be able to use Querydsl at all.

## Conclusion

It seems that we can enhance the compatibility of APT. Try to recognize more annotations, or even add our own annotations to provide metadata. This method is suitable for both maven and gradle, even kotlin.

I know it's a big project. I'm happy to help, but I don't know how to start it.

For example, I haven't figured out what the nature of APT is? Just generate Q class? Maybe not. Because I found that the result of`apt:jpa` and `apt:general` are both the same -- extend `EntityPathBase<T>`, But it can be used to update in JPA, but not in querydsl-sql. What makes this difference? 

Do other people have any opinions? I am very interested in the development of this project.

**Comment by jwgmeligmeyling** [comment]
> but it actually contains many outdated content.

It's actually not too bad, but it can be cleaned from parts that are just not used. Who uses Spring Roo, Ant, legacy Hibernate hbm mapping files these days? The first chapters bothers you with all these concepts. Its also too easily to scroll from `querydsl-jpa` to `querydsl-jdo` or even `querydsl-sql` documentation while they are completely different use cases. So agreed, the documentation needs to be improved. And the easiest way is to get rid of the junk at least 😉 

>  I want to help improve it, but the document generator itself is old too

I've read up a bit on it, its not too bad. Also considering the fact that we want to generate a site from the documentation as well.

> 2. Too many bindings with JPA

I have to disagree on this one. It has bindings with Java. That means it works well with query languages that work in the Java domain. Like: JDO, JPA and `querydsl-collections` (which queries in-memory objects).

There is no need for you to explain why you dislike JPA. `querydsl-jpa` is a JPA library. `querydsl-jpa` is by far the most popular Querydsl module. If you don't want to use `querydsl-jpa`, just don't use it. `querydsl-sql` is by no means crippled by the fact that `querydsl-jpa` exists. `querydsl-sql` is crippled by the fact that, contrary to all the other Querydsl modules, it relies on the static metamodel not just for the expression types but for the entire query generation as well. Because for JPA and JDO and others, this is relayed to the ORM provider instead. Querydsl is not an ORM, and that is why `querydsl-sql` is intrinsically so much harder to do right.

Personally I have little interest in `querydsl-sql`. I am not using it in any of my projects. Those tend to move to other types of storage than just relational storage anyway. Querydsl performs best in massive projects that already have hundreds of database tables and entities already, and that are desperately seeking a tool that allows for type safe query generation and most importantly, an exploratory programming style through auto completion. These projects almost always already run JPA or JDO, and Querydsl performs really well though. For SQL however, I am not sure if we have a real advantage over newer tools that entered the game, like Spring JDBC or JOOQ. If you believe so, by all means go ahead and fix the issues. Most requested indeed is a Gradle counter part of the Maven plugin for SQL metamodel generation.

> 3. Q class code gen is not consistent

1. You don't need codegen for Querydsl to work. 
  * It works just fine with a manually programmed metamodel. Some projects even prefer this for performance reasons, and it works well for them.
  * There is also `Alias.*` and others that create a static metamodel at runtime. 
2. Codegen is consistent for all but one module of Querydsl, the only exception being `querydsl-sql` which needs to be able to resolve the table names and columns etc.

> First of all, APT can be used with gradle, or even simpler than maven. But only for JPA (see problem 3).

There are plenty of examples that use APT just fine with Gradle, and also with KAPT for Kotlin. Its just the Maven plugin that can be used for `querydsl-sql` that lacks a Gradle counter part.

Now if we can get rid of the Maven plugin by creating an APT plugin for `querydsl-sql` and get Gradle support essentially for free, I am all up for it.

FWIW, `apt` and `general` differ from the set of annotations they use for metamodel generation. Output might be the same depending on how the entities are annotated. AFAIK entities generated using the `general` plugin (or the `jpa` plugin for that matter, although having to use JPA annotations in this use case would be inconvenient) can be used with `querydsl-sql` just fine, as long as they use the proper entity naming convention and a properly configured instance for `Configuration` is provided on runtime.

**Comment by natros** [comment]
I have a gradle plugin that I use privately and fulfills my needs. I can make this plugin available, but I don't have time to maintain it.

An example of use:

```gradle
plugins {
  id 'local.java-library'
  id 'com.github.natros.querydsl'
}

querydsl {
  template.set('ORACLE')
  jdbcDriver = rootProject.ext.jdbcDriver
  jdbcUrl = rootProject.ext.jdbcUrl
  jdbcUser = rootProject.ext.jdbcUser
  jdbcPassword = rootProject.ext.jdbcPassword
  schema.set('RFID')
  packageName.set('pt.ipb.rfid.controller.entities')
  namingStrategyClass.set('com.github.natros.wf.querydsl.sql.codegen.TPrefixTableNamingStrategy')
  beanSerializerClass.set('com.github.natros.wf.querydsl.codegen.GuavaBeanSerializer')
  // exportForeignKeys.set(false)
  typeMappings {
    AGENDA_CURSO {
      column.ANO = 'java.lang.Integer'
      column.AGENDA_CURSO_ID = 'java.util.UUID'
      column.REPETICAO = 'pt.ipb.rfid.controller.shared.RepeticaoEnumType'
    }
    AGENDA_GRUPO {
      column.AGENDA_GRUPO_ID = 'java.util.UUID'
      column.REPETICAO = 'pt.ipb.rfid.controller.shared.RepeticaoEnumType'
    }
    BOXLOG {
      column.RFID = 'java.lang.Long'
    }
    CARTAO_RFID {
      column.RFID = 'java.lang.Long'
    }
    IPB_CURSOS {
      column.DURACAO = 'java.lang.Integer'
    }
    IPB_LECT_INI {
      column.ANO_CURRICULAR = 'java.lang.Integer'
      column.TIPO_MAT = 'java.lang.Integer'
    }
    PICAGENS {
      column.RFID = 'java.lang.Long'
      column.CARTAO_ACTIVO = 'java.lang.Boolean'
    }
    SALA {
      column.TIPO_TERMINAL = 'pt.ipb.rfid.controller.shared.TipoTerminalEnumType'
    }
  }
}

dependencies {
  api "com.querydsl:querydsl-core:$querydslVersion"
  api "com.querydsl:querydsl-sql:$querydslVersion"

  implementation project(':rfid-controller-shared')
  implementation "com.google.guava:guava:$guavaVersion"
  implementation 'jakarta.validation:jakarta.validation-api:2.0.2'

  compileOnly 'jakarta.annotation:jakarta.annotation-api:1.3.5'

  querydsl "com.oracle.database.jdbc:ojdbc11:$ojdbVersion"
  querydsl "com.github.natros.wf:wf-querydsl-sql-naming:$wfVersion"
  querydsl "com.github.natros.wf:wf-querydsl-codegen-guava:$wfVersion"
}
```

**Reply by jwgmeligmeyling** [reply]
If you would contribute the gradle plugin as module in a PR, I would definitely be willing to support it as far as merging future PRs, and maintaining binary compatibility and ensuring the tests keep passing with further developments of Querydsl.

**Comment by ghost** [comment]
I think we need a new document or guide, maybe it already exists, but I didn't find it.
In this [discussion](https://github.com/querydsl/querydsl/discussions/2936#discussioncomment-908629) I gained some information, but not more, which made me feel pain.
